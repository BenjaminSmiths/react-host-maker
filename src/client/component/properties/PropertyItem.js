import React, {Component} from 'react';
import PropTypes from 'prop-types';

const FormatUtils = {

    capFirst: (name) => {
        return name.charAt(0).toUpperCase() + name.slice(1);
    },

    buildAddress: ({line1, line2, line3, line4, city, country, postCode}) => {
        let house = !line2 ? [line1] : [`${line1}, ${line2}`];
        let streets = [line3, line4].filter((l) => !!l);
        let other = [`${city}, ${country}`, postCode];

        return [...house, ...streets, ...other]
            .map((val, key) => {
                return (
                    <div key={key}>{val}</div>
                );
            });
    },

    // to be replaced with a proper multi currency format
    // but it will do for now.
    formatIncome: (x, country) => {
        x = x.toString();
        let pattern = /(-?\d+)(\d{3})/;
        while (pattern.test(x)) {
            x = x.replace(pattern, "$1, $2");
        }
        return x;
    }
};

class PropertyItem extends Component {

    render() {
        let {property} = this.props;

        return (
            <div>
                <div className="owner">{FormatUtils.capFirst(property.owner)}</div>
                <div className="address">
                    {
                        FormatUtils.buildAddress(property.address)
                    }
                </div>
                <div className="income">&pound;{FormatUtils.formatIncome(property.incomeGenerated)}</div>
            </div>
        )
    }
}

PropertyItem.propTypes = {
    property: PropTypes.object.isRequired
};

export default PropertyItem;