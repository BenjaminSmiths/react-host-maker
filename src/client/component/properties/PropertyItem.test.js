import React from 'react';
import ReactDOM from 'react-dom';
import PropertyItem from './PropertyItem';
import ReactTestUtils from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

describe('PropertyItem', () => {
    let property = {
        'owner': 'carlos',
        'address': {
            'line1': 'Flat 5',
            'line4': '7 Westbourne Terrace',
            'postCode': 'W2 3UL',
            'city': 'London',
            'country': 'U.K.'
        },
        'incomeGenerated': 2000.34,
        'image': 'property1.jpg'
    };

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<PropertyItem property={property}/>, div);
    });

    it('should find the owner text', () => {

        let component = ReactTestUtils.renderIntoDocument(
            <PropertyItem property={property}/>
        );

        let owner = ReactTestUtils.findRenderedDOMComponentWithClass(
            component, 'owner'
        );

        expect(ReactDOM.findDOMNode(owner).textContent).toEqual('Carlos');
    });

    it('should find the income amount', () => {

        let component = ReactTestUtils.renderIntoDocument(
            <PropertyItem property={property}/>
        );

        let income = ReactTestUtils.findRenderedDOMComponentWithClass(
            component, 'income'
        );

        expect(ReactDOM.findDOMNode(income).textContent).toEqual('Â£2, 000.34');
    });

    describe('Format Address', () => {

        it('should find the address block', () => {

            let component = ReactTestUtils.renderIntoDocument(
                <PropertyItem property={property}/>
            );

            let address = ReactTestUtils.findRenderedDOMComponentWithClass(
                component, 'address'
            );

            expect(ReactDOM.findDOMNode(address)).toBeDefined()
        });

        it('should match the property missing line2', () => {
            const component = renderer.create(<PropertyItem property={{
                'owner': 'carlos',
                'address': {
                    'line1': 'Flat 5',
                    'line4': '7 Westbourne Terrace',
                    'postCode': 'W2 3UL',
                    'city': 'London',
                    'country': 'U.K.'
                },
                'incomeGenerated': 2000.34
            }}/>);
            expect(component.toJSON()).toMatchSnapshot();
        });

        it('should match the property with all lines', () => {
            const component = renderer.create(<PropertyItem property={{
                'owner': 'ankur',
                'address': {
                    'line1': '4',
                    'line2': 'Tower Mansions',
                    'line3': 'Off Station road',
                    'line4': '86-87 Grange Road',
                    'postCode': 'SE1 3BW',
                    'city': 'London',
                    'country': 'U.K.'
                },
                'incomeGenerated': 1000
            }}/>);
            expect(component.toJSON()).toMatchSnapshot();
        });

        it('should match the property line3 missing', () => {
            const component = renderer.create(<PropertyItem property={{
                'owner': 'elaine',
                'address': {
                    'line1': '4',
                    'line2': '332b',
                    'line4': 'Goswell Road',
                    'postCode': 'EC1V 7LQ',
                    'city': 'London',
                    'country': 'U.K.'
                },
                'incomeGenerated': 1200
            }}/>);
            expect(component.toJSON()).toMatchSnapshot();
        });

    })
});