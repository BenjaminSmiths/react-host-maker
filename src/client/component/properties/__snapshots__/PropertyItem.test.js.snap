// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyItem Format Address should match the property line3 missing 1`] = `
<div>
  <div
    className="owner"
  >
    Elaine
  </div>
  <div
    className="address"
  >
    <div>
      4, 332b
    </div>
    <div>
      Goswell Road
    </div>
    <div>
      London, U.K.
    </div>
    <div>
      EC1V 7LQ
    </div>
  </div>
  <div
    className="income"
  >
    £
    1, 200
  </div>
</div>
`;

exports[`PropertyItem Format Address should match the property missing line2 1`] = `
<div>
  <div
    className="owner"
  >
    Carlos
  </div>
  <div
    className="address"
  >
    <div>
      Flat 5
    </div>
    <div>
      7 Westbourne Terrace
    </div>
    <div>
      London, U.K.
    </div>
    <div>
      W2 3UL
    </div>
  </div>
  <div
    className="income"
  >
    £
    2, 000.34
  </div>
</div>
`;

exports[`PropertyItem Format Address should match the property with all lines 1`] = `
<div>
  <div
    className="owner"
  >
    Ankur
  </div>
  <div
    className="address"
  >
    <div>
      4, Tower Mansions
    </div>
    <div>
      Off Station road
    </div>
    <div>
      86-87 Grange Road
    </div>
    <div>
      London, U.K.
    </div>
    <div>
      SE1 3BW
    </div>
  </div>
  <div
    className="income"
  >
    £
    1, 000
  </div>
</div>
`;
